sdk/core/Azure.Core/src/DynamicData/MutableJsonDocument.ChangeTracker.cs

```diff
@@ -135,6 +141,70 @@ internal IEnumerable<MutableJsonChange> GetRemovedProperties(string path, int hi
                 }
             }
 
+            internal MutableJsonChange? GetFirstMergePatchChange(ReadOnlySpan<char> rootPath, out int maxPathLength)
+            {
+                // This method gets the first change from the list in sorted order by path
+                // It also returns the max path length of changes on the list.
+
+                maxPathLength = -1;
+
+                if (_changes == null)
+                {
+                    return null;
+                }
+
+                MutableJsonChange? min = null;
+
+                for (int i = _changes!.Count - 1; i >= 0; i--)
+                {
+                    MutableJsonChange c = _changes[i];
+
+                    if (c.Path.AsSpan().StartsWith(rootPath) &&
+                        (min == null || c.IsLessThan(min.Value.Path.AsSpan())))
+                    {
+                        min = c;
+                    }
+
+                    if (c.Path.Length > maxPathLength)
+                    {
+                        maxPathLength = c.Path.Length;
+                    }
+                }
+
+                return min;
+            }
+
+            internal MutableJsonChange? GetNextMergePatchChange(ReadOnlySpan<char> rootPath, ReadOnlySpan<char> lastChangePath)
```

Comment:
Any value in implementing an enumerator for patch changes since that you don't need to start over from first and each subsequent call to get next changes? Seems like state could be retained (in the enumerator) to avoid potentially extraneous iterations for all `_changes` each time.

Reply:
I was concerned that an enumerator would make unwanted allocations, but if you know a way to avoid that, I would be interested to learn more.

Reply:
Use a struct enumerator. That's what much (maybe even most) of the BCL enumerators are. Were there other allocations you were worried about?

Reply:
Is there an example in dotnet/runtime you would suggest modeling one after?

Reply:
you can create an Enumerator struct, add MoveNext method and Current property, and you will be able to foreach over this struct, i.e. no need to implement any interfaces (that would cause boxing). See https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L1145

Reply:
Not in particular. Many - like [`List<T>.Enumerator`](https://github.com/dotnet/runtime/blob/910782efdd6ab1cb61f00c71fd9e4f2bff32ba0c/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L1145) - are structs, but here you could model it fairly easily, it seems, incomplete e.g.,

```c#
struct Enumerator : IEnumerator<MutableJsonChange>, IEnumerator
{
  private readonly MutableJsonDocument _doc;
  private readonly ReadOnlySpan<char> _root;
  public Enumerator(MutableJsonDocument doc, ReadOnlySpan<char> root)
  {
    _doc = doc;
    _root = root;
  }
  public MutableJsonChange Current { get; private set; }
  object IEnumerator.Current => Current;
  public bool MoveNext()
  {
    if (_doc._changes is null) return false;
    /** do the rest of the stuff and store each match to Current **/
  }
}
```

Reply:
True, you don't need to implement any interfaces, but the .NET team optimized the boxing out ages ago e.g., https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgIwFgBQyAMABMugHQDCA9hBAKYDGwYFAdgM4DcO+R6ALJ7gDMRVATIEQPQQB4wzYAD4cAbxwF1BAA4AnMADcAhsBoFtNAzBYQAngTnAA2gF0CAfQPbtBALwFmNAHdnAmUCPCgCdAjUCOEAXwENAjUNYgA2Il4CAFkDOQAKAEoU9VVsJKSAMwozAzoACwJ8+wIaWgBbO2Y/QLEi4vKK0pKh1PQATny2mnbCxNG4kcXB9RHkYQBJAFFmAFd2mm0jGtl5BQIAcRpgHf3D4+0in3P/AIJbg6PgE/sFfPdPHMRts9p8HlISFcbqD7t9HoVnpdrh9YTUigIRqxgNpdgx3jCvicACrnSQgu6E7TSEkRclguEjMqjLS6QzGUzmSzMGwEInBAHaeZDHT6IwmFquCiVSqsa4+MJCirrfEUh75PkuDzaBGhAXyrXsAjLZnKoliXaeGjyEIEADm10NIrZJllwENyxGSQoACMAFb0YAEOmo7TkC1ma3ec5kcNWt04T2pYTILIAETArE0FFlT2UHpWSuE3ooVByFD0NAAcjQAB7AIqJ4YF5lgSpNSXS10EaRuLUkAAyVttwHqA2ZSSZ46GMctkd7ngcHZl13g8Cciqn6mQAHYCNjdjQNwtG0rd5UDBBZUf1PmTcmsgAlGiu3PGjTLOJAA=. We had to do this for a lot of the VS start up code in the VS setup engine. The only boxing is if you use `IEnumerator.Current` instead of `IEnumerator<T>.Current`. That's IIF the compiler knows of your specific `Enumerator` implementation, which in this case - whether you implement the interfaces or not (which makes them more useful elsewhere, if/when needed) - it would.

Reply:
I added one to iterate over segments (e.g. to implement `Split`), but still working on the one for changes separately - there was some magic that merged array changes into a single change that I haven't figured out how to address yet.

Reply:
If the goal of introducing the enumerator for this case is to avoid iterating over items in the change list, I don't think that's possible given the need to retrieve items from the list in sorted order (this implementation replaced a less efficient call to a list sort, which wasn't a stable sort anyway).  Since there is logic in GetNext that takes state from the loop, I would prefer to keep the current implementation, unless there is another goal I am missing.

Reply:
It was to avoid what seemed like an O(N*M) call - looping through the same set of changes (N) with each call itself in a loop (M). A stateful enumerator could reduce that to O(N) if the changes were already sorted. If M is small, maybe it's not worth it.

Reply:
Yeah, I think the idea is that we expect the number of changes to be relatively small.  N is every change on the list, and M is every unique changed element, removing elements whose ancestors changed, and any changes to array elements.  If it turns out to be a problematic bottleneck, we can revisit for sure.